### PR TITLE
Added new payment request for deferred payments

### PR DESCRIPTION
**Pull request by Rattler3**
It is currently not possible to create a new Sage pay transaction of any type other then PAYMENT when using server integration, this adds the functionality for DEFERRED payments.

### CODE REVIEW

**Comment by Rattler3** [comment]
@judgej can you review this PR?

**Comment by judgej** [comment]
Quick question: is a deferred purchase request any different to a authorise request? If not, then the `ServerDeferredPurchaseRequest` is already covered by the `ServerAuthorizeRequest`. They look the same to me (in the code), but I may be missing something.

The `CaptureRequest` is then used to capture what has been authorised (aka DEFERRED).

This is a serious question, because I realise there are two different ways of deferring payment in Sage Pay, each involving different ways of collecting (e.g. one much be collected in full just one, and the other can be an open authorisation that can have multiple amounts up to a limit be collected over a longer period of time). TBH, I'm not fully up to speed on those features, as they can be hard to follow through the docs.

**Comment by judgej** [comment]
Okay, looked the docs and the code.

At the moment, this gateway driver implements `Authorize` requests using the `DEFERRED` payment type. So on the face of it, the payment type your PR implements is already supported.

A bit of background on the payment types that Sage Pay supports:

* PAYMENT - register, authorise and settle the transaction in one go.
* DEFERRED - register and authorise the payment, but settle it (RELEASE aka Capture) at a later date, within 30 days (ideally stick closer to 6 days).
* AUTHENTICATE - authenticate the card and user details, but do not authorise payment. These details are held for 90 days (with some exceptions) and can be authorised later without the user needing to be present again. The result is not determined by funds available; that is done later.
* AUTHORISE - authorise a payment against a previously authenticated card. Value can be up to 115% of the authenticated amount. User does not need to be present. Whether the authorisation is successful, will depend on funds available at this time.

Use PAYMENT and DEFERRED in most instances (both supported by OmniPay). Use AUTHENTICATE+AUTHORISE if the final amount is not fully known up front, or authorisation or release cannot normally be done within a reasonable time (6 days usually, or 30 days absolute max).

**Comment by judgej** [comment]
Conclusion: OmniPay already supports DEFERRED payments, and this PR is not needed. But please kick me if I've misunderstood the whole thing.

**Comment by Rattler3** [comment]
Thanks for looking into this and doing alot of reading/research.

The reason I created this, is that when using the ```ServerPurchaseRequest``` the action is already set to PAYMENT this meant that the payment was taken straight away and not DEFERRED. I created the new ```ServerDeferredPurchaseRequest``` to allow DEFERRED payments to be created.

Using the existing ```ServerPurchaseRequest``` how would you create a DEFERRED payment?

**Comment by judgej** [comment]
`ServerPurchaseRequest` is to take a payment. If you just want to authorise and take the payment later, then the `ServerAuthorizeRequest` and `CaptureRequest` messages serve that purpose.

**Comment by judgej** [comment]
~~~php
$gateway = OmniPay::create('SagePay\Server');
...
// Use "authorize" here and not "purchase"
$requestMessage = $gateway->authorize([...]);
$responseMessage = $requestMessage->send();
~~~

**Comment by Rattler3** [comment]
I see, feel free to close this PR as is no longer needed.

**Comment by judgej** [comment]
Will do, thanks. Let us know how it goes, switching over from your branch.
